suprb/optimizer/rule/mutation.py

```diff
@@ -69,47 +87,108 @@ class Normal(RuleMutation):
+        assert isinstance(self.sigma, np.ndarray) and self.sigma.shape[0] == 2
+        bounds = rule.match.bounds
+        bounds[:, 0] += random_state.normal(scale=self.sigma[0],
+                                            size=rule.match.bounds.shape[0])
+        bounds[:, 1] += random_state.normal(scale=self.sigma[1],
+                                            size=rule.match.bounds.shape[0])
+
 
 class Halfnorm(RuleMutation):
     """Sample with (half)normal distribution around the center."""
 
-    def mutation(self, dimensions: int, random_state: RandomState):
-        return halfnorm.rvs(scale=self.sigma / 2, size=dimensions,
-                            random_state=random_state)
-
     def ordered_bound(self, rule: Rule, random_state: RandomState):
         bounds = rule.match.bounds
         mean = np.mean(bounds, axis=1)
-        bounds[:, 0] = mean - self.mutation(dimensions=bounds.shape[0], random_state=random_state)
-        bounds[:, 1] = mean + self.mutation(dimensions=bounds.shape[0], random_state=random_state)
+        bounds[:, 0] = mean - \
+            halfnorm.rvs(scale=self.sigma / 2, size=bounds.shape[0], random_state=random_state)
+
+        bounds[:, 1] = mean + \
+            halfnorm.rvs(scale=self.sigma / 2, size=bounds.shape[0], random_state=random_state)
         rule.match.bounds = np.sort(rule.match.bounds, axis=1)
 
+    def unordered_bound(self, rule: Rule, random_state: RandomState):
+        bounds = rule.match.bounds
+        mean = np.mean(bounds, axis=1)
+        bounds[:, 0] = mean - \
+            halfnorm.rvs(scale=self.sigma / 2, size=bounds.shape[0], random_state=random_state)
+
+        bounds[:, 1] = mean + \
+            halfnorm.rvs(scale=self.sigma / 2, size=bounds.shape[0], random_state=random_state)
+
+    def centre_spread(self, rule: Rule, random_state: RandomState):
+        raise TypeError("Halform Mutation is not implemented for CSR")
+
+    def min_percentage(self, rule: Rule, random_state: RandomState):
+        raise TypeError("Halform Mutation is not implemented for MPR")
+
 
 class HalfnormIncrease(RuleMutation):
     """Increase bounds with (half)normal noise."""
 
-    def mutation(self, dimensions: int, random_state: RandomState):
-        return halfnorm.rvs(scale=self.sigma / 2, size=dimensions,
-                            random_state=random_state)
-
     def ordered_bound(self, rule: Rule, random_state: RandomState):
         bounds = rule.match.bounds
-        bounds[:, 0] -= self.mutation(dimensions=bounds.shape[0], random_state=random_state)
-        bounds[:, 1] += self.mutation(dimensions=bounds.shape[0], random_state=random_state)
+        bounds[:, 0] -= halfnorm.rvs(scale=self.sigma / 2, size=bounds.shape[0], random_state=random_state)
+        bounds[:, 1] += halfnorm.rvs(scale=self.sigma / 2, size=bounds.shape[0], random_state=random_state)
         rule.match.bounds = np.sort(rule.match.bounds, axis=1)
```

Comment:
Please don't just resolve conversations, but also answer the question

Reply:
For both OBR and UBR only one sigma is required/available whereas for CSR and MPR two sigmas are used. Because of this, sigma is either a float value or an array and thus used depending on the representation.